### ISSUE TITLE
Can't create reverse route through lanelets that are not oneway

### ISSUE THREAD

**Issue by kgreenek** [fzi-forschungszentrum-informatik/Lanelet2#77]
I am attempting to label sidewalks as lanelets, and create routes through the sidewalks using pedestrian traffic rules.

It works great and was very easy to get started. However, I noticed that when I try to create routes that go in the reverse direction from how the lanelets were created, the call to graph->shortestPath() returns an empty optional.

My lanelets were created with `one_way = no` in JOSM.

Am I interpreting that field correctly?

In short, say you have 3 lanelets `A -> B -> C`. I can plan a route successfully from A to C, but trying to plan a route from C to A fails even though the lanelets are marked with one_way = no.

**Comment by poggenhans** [comment]
I suppose it is because you didn't invert the lanelets for your query. If you search for a path from C to A, the result is empty because there actually exists no path that allows you to go *forward* and reach A. If you search a path from `C.invert()` to `A.invert()`, it should work.

**Comment by kgreenek** [comment]
Thanks for the feedback! Here is a work-around I came up with based on your comment:

```c++
boost::optional<lanelet::routing::LaneletPath> ShortestPath(
    const lanelet::routing::RoutingGraph& graph, const lanelet::Lanelet& from,
    const lanelet::Lanelet& to) {
  const boost::optional<lanelet::routing::LaneletPath> paths[] = {
      graph.shortestPath(from, to),
      graph.shortestPath(from.invert(), to),
      graph.shortestPath(from, to.invert()),
      graph.shortestPath(from.invert(), to.invert()),
  };
  boost::optional<lanelet::routing::LaneletPath> shortest_path;
  for (const auto& path : paths) {
    if (path) {
      if (!shortest_path || path->size() < shortest_path->size()) {
        shortest_path = path;
      }
    }
  }
  return shortest_path;
}
```

This works! However, it's probably doing extra work that it doesn't need to, and the shortest path will be the one with the shortest number of lanelets out of those 4 combinations, instead of the shortest distance in meters.

It feels a klunky to have to check forwards and backwards in this case, since sidewalks inherently do not have a direction (at least how I would like to use them).

How would you feel about modifying the graph search logic to add in those edges automatically if the "from" lanelet or "to" lanelet have one_way set to false?

**Comment by kgreenek** [comment]
I think it would also be more consistent that way. Right now B is allowed to be traversed forward or backwards because it has one_way set to false, but A and C are treated differently.

**Comment by poggenhans** [comment]
That might seem logical from a pedestrian point of view, because they can turn on the spot. For vehicles, it makes a huge difference if a route starts in a certain direction or the other way round.

If you really don't care about the orientation, the solution you came up with seems to be the simplest way to do it.

One small side note: Better use `lanelet::Optional` to refer to optional types returned by lanelet2 instead of `boost::optional`. One day, lanelet2 might switch to using `std::optional` and your code will break.

**Comment by kgreenek** [comment]
Circling back on this. That makes sense for a car that is already driving on the road. You wouldn't want to give it directions that require it to immediately make a U-turn.

I do still argue that it is odd, even in the "car on road" use-case, that the destination lanelet can't be inverted (assuming the destination lanelet is not one-way). If the destination isn't one-way, then it is valid to reach it from either direction.

Another use-case that comes to mind is planning a route from a start lanelet to a destination lanelet, when you aren't already on the road. For example, the car might be parked in a garage, and want to know the fastest route from the lanelet nearest the garage to some destination address's lanelet. In that case, it wouldn't matter if the car turns right or left when leaving the garage. I'm not sure the best solution to this case; perhaps calling the shortestPath method twice is the best solution for that.

**Comment by poggenhans** [comment]
> I do still argue that it is odd, even in the "car on road" use-case, that the destination lanelet can't be inverted (assuming the destination lanelet is not one-way). If the destination isn't one-way, then it is valid to reach it from either direction.

I can see that this behaviour may be unintuitive. But there might be people who care about orientation at the destination and who find the other behaviour unintuitive. Therefore I advocate the more fine-grained solution.

> Another use-case that comes to mind is planning a route from a start lanelet to a destination lanelet, when you aren't already on the road. For example, the car might be parked in a garage, and want to know the fastest route from the lanelet nearest the garage to some destination address's lanelet.

In the concept of lanelet2, a lanelet is the "I care about orientation here" geometry. The "i don't care about orientation here" geometry would be an *area*. Therefore it could make sense to make the parking spot of the vehicle an area.

There is a slight overhead if you have to query shortestPath function twice in order to find the actually best path, but since that computation is quite fast (and you only need it for bidirectional lanelets).

**Comment by poggenhans** [comment]
This issue has been inactive for a while. I'm assuming the original problem has been resolved.

**Comment by dufengit** [comment]
have you ever tried   appendBidirectionalLanelets(passableLanelets) for reverse lanelets ?
